src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregatorStore.cs

```diff
@@ -0,0 +1,472 @@
+        // FixedSizeLabelNameDictionary<StringSequenceMany, ConcurrentDictionary<ObjectSequenceMany, TAggregator>>
+        // MultiSizeLabelNameDictionary<TAggregator> - this is used when we need to store more than one of the above union items
+        private volatile object? _stateUnion;
+        private volatile AggregatorLookupFunc<TAggregator>? _cachedLookupFunc;
+        private Func<TAggregator> _createAggregatorFunc;
+
+        public AggregatorStore(Func<TAggregator> createAggregator)
+        {
+            _stateUnion = default;
+            _cachedLookupFunc = default;
+            _createAggregatorFunc = createAggregator;
+        }
+
+        public TAggregator GetAggregator(ReadOnlySpan<KeyValuePair<string, object?>> labels)
+        {
+            AggregatorLookupFunc<TAggregator>? lookupFunc = _cachedLookupFunc;
+            if (lookupFunc != null)
+            {
+                if (lookupFunc(labels, out TAggregator? aggregator)) return aggregator!;
+            }
+
+            // slow path, label names have changed from what the lookupFunc cached so we need to
+            // rebuild it
+            return GetAggregatorSlow(labels);
+        }
+
+        private TAggregator GetAggregatorSlow(ReadOnlySpan<KeyValuePair<string, object?>> labels)
+        {
+            AggregatorLookupFunc<TAggregator> lookupFunc = LabelInstructionCompiler.Create(ref this, _createAggregatorFunc, labels);
+            _cachedLookupFunc = lookupFunc;
+            bool match = lookupFunc(labels, out TAggregator? aggregator);
+            Debug.Assert(match);
+            return aggregator!;
+        }
+
+        public void Collect(Action<LabeledAggregationStatistics> visitFunc)
+        {
+            object? stateUnion = _stateUnion;
+            if (stateUnion is TAggregator agg)
+            {
+                IAggregationStatistics stats = agg.Collect();
+                visitFunc(new LabeledAggregationStatistics(stats));
+            }
+            else if (stateUnion is FixedSizeLabelNameDictionary<StringSequence1, ObjectSequence1, TAggregator> aggs1)
+            {
+                aggs1.Collect(visitFunc);
+            }
+            else if (stateUnion is FixedSizeLabelNameDictionary<StringSequence2, ObjectSequence2, TAggregator> aggs2)
+            {
+                aggs2.Collect(visitFunc);
+            }
+            else if (stateUnion is FixedSizeLabelNameDictionary<StringSequence3, ObjectSequence3, TAggregator> aggs3)
+            {
+                aggs3.Collect(visitFunc);
+            }
+            else if (stateUnion is FixedSizeLabelNameDictionary<StringSequenceMany, ObjectSequenceMany, TAggregator> aggsMany)
+            {
+                aggsMany.Collect(visitFunc);
+            }
+            else if (stateUnion is MultiSizeLabelNameDictionary<TAggregator> aggsMultiSize)
+            {
+                aggsMultiSize.Collect(visitFunc);
+            }
```

Comment:
Beauty is in the eye of the beholder, but if you want, you can rewrite this as a pattern matching switch:
```C#
switch (_stateUnion)
{
    case TAggregator agg:
        IAggregationStatistics stats = agg.Collect();
        visitFunc(new LabeledAggregationStatistics(stats));
        break;

    case FixedSizeLabelNameDictionary<StringSequence1, ObjectSequence1, TAggregator> aggs1:
        aggs1.Collect(visitFunc);
        break;

    case FixedSizeLabelNameDictionary<StringSequence2, ObjectSequence2, TAggregator> aggs2:
        aggs2.Collect(visitFunc);
        break;

    ...
}
```
I personally prefer the switch style, but up to you.

Reply:
I'm happy to match team preference : )